core/include/detray/core/detail/container_buffers.hpp

```diff
@@ -71,38 +72,37 @@ concept device_buffer =
 namespace detail {
 /// Helper trait to check whether a type has a [vecmem] buffer type defined
 /// @{
-template <class T>
-struct get_buffer : public std::false_type {
-    using type = void;
-};
+template <class T, typename = void>
+struct get_buffer {};
 
 template <class T>
-requires concepts::device_buffer<typename T::buffer_type> struct get_buffer<T>
-    : public std::true_type {
+struct get_buffer<
+    T, std::enable_if_t<
+           std::is_class_v<typename std::remove_cv_t<T>::buffer_type>, void>> {
```

Comment:
```suggestion
           detray::detail::is_buffer_v<typename std::remove_cv_t<T>::buffer_type>, void>> {
```

Reply:
😆 I was sure that you would bring this up.

`detray::details::is_buffer_v`, contrary to its name, uses concepts internally itself. As such, doesn't work with GCC 12.

But at the same time, this is a bit of a chicken-and-egg situation here. We use this trait (`detray::concepts::detail::get_buffer`) **only** to check whether a given type "is bufferable". The type returned by this trait is always checked for whether it is a buffer.

This trait on the other hand, only advertises that it would get "the buffer type" for a given type. So I decided that if something has a `buffer_type` member, that would just be used as "the buffer type" of that type. So that the code calling this, could check if it's a "good buffer type". See:
  - https://github.com/acts-project/detray/blob/main/core/include/detray/core/detail/container_buffers.hpp#L64-L67
  - https://github.com/acts-project/detray/blob/main/core/include/detray/core/detail/container_buffers.hpp#L118

Reply:
I am not really following (maybe I need to see the error). Where does ```detray::detail::is_buffer``` use a concept? It is just a trait, which is needed to tag vecmem buffer types as buffers in detray, since I cannot make them inherit from the detray buffer tag (```dbase_buffer```). But it is probably correct, that ```is_buffer``` does not work anymore for ```get_buffer```, since we need to teach it again how to recognize detray internal buffers. I had intended to only use it for vecmem types in the background at this point and exclusively use the concept ```detray::concepts::device_buffer``` everywhere else, but if the concept is not understood by some compilers, we alas need to fall back on ```is_buffer``` again. This kind of makes the actual concept  a bit redundant, but I would still like to keep it around everywhere else, because when all relevant compilers can handle concepts correctly, we should remove the ```enable_if``` again imo.

```detray::detail::get_buffer_t``` is definitely used in a few key places outside of this header to extract the buffer type of classes and not just for the ```bufferable``` concept. In general, I think it would be great, if it could check not just that ```buffer_type``` exists, but that it is actually a buffer type. This is meant to make the code easier to debug, instead of just inserting whatever it finds in ```buffer_type```. See e.g.
https://github.com/acts-project/detray/blob/2d47f73241091fc8cc1f6e36050a8a8fd69489db/core/include/detray/utils/grid/grid_collection.hpp#L131

Reply:
I should probably have renamed it to ```is_vecmem_buffer``` after introducing the concept. But I guess, smth like this should make it a proper ``` is_buffer``` trait again:
```
template <typename T>
struct is_buffer<
    T, std::enable_if_t<
           std::is_base_of_v<detray::detail::dbase_buffer,
                             std::remove_cvref_t<T>>,
           void>> : public std::true_type {};
```
It is funny though, that the very same pattern works with views...

Reply:
I.e. this https://github.com/acts-project/detray/blob/2d47f73241091fc8cc1f6e36050a8a8fd69489db/core/include/detray/core/detail/container_views.hpp#L96 compiles fine?

Reply:
At least no part of the Detray or traccc build was complaining about that code. I imagine that that code is being tested as well, but maybe not through quite the same kind of call chain. 🤔

Reply:
There are e.g. no overloads for ```const``` buffer types in the detray type traits, because I was under the impression they would not be used... But I don't understand why this compiler version would make a difference there

Reply:
So... I make the following modification to my PR's code locally:

```diff
diff --git a/core/include/detray/core/detail/container_buffers.hpp b/core/include/detray/core/detail/container_buffers.hpp
index 160dd053..aabf8695 100644
--- a/core/include/detray/core/detail/container_buffers.hpp
+++ b/core/include/detray/core/detail/container_buffers.hpp
@@ -77,8 +77,9 @@ struct get_buffer {};
 
 template <class T>
 struct get_buffer<
-    T, std::enable_if_t<
-           std::is_class_v<typename std::remove_cv_t<T>::buffer_type>, void>> {
+    T, std::enable_if_t<detray::detail::is_buffer_v<
+                            typename std::remove_cv_t<T>::buffer_type>,
+                        void>> {
     using type = typename T::buffer_type;
 };
```

And this results in a horrificly long error message, with this being "at its core":

```
...
/home/krasznaa/ATLAS/projects/detray/detray/core/include/detray/core/detail/container_buffers.hpp:119:9:   required for the satisfaction of ‘bufferable<G>’ [with G = detray::grid_impl<detray::axis::multi_axis<false, detray::polar2D<detray::cmath<float> >, detray::axis::single_axis<detray::axis::closed<detray::axis::label::e_x>, detray::axis::regular<detray::container_types<vecmem::vector, detray::tuple, std::array, vecmem::jagged_vector, detray::dmap>, float> >, detray::axis::single_axis<detray::axis::circular<detray::axis::label::e_y>, detray::axis::regular<detray::container_types<vecmem::vector, detray::tuple, std::array, vecmem::jagged_vector, detray::dmap>, float> > >, detray::bins::single<detray::material_slab<float> >, detray::simple_serializer>]
/home/krasznaa/ATLAS/projects/detray/detray/core/include/detray/utils/grid/detail/concepts.hpp:72:9:   required for the satisfaction of ‘grid<grid_t>’ [with grid_t = detray::grid_impl<detray::axis::multi_axis<false, detray::polar2D<detray::cmath<float> >, detray::axis::single_axis<detray::axis::closed<detray::axis::label::e_x>, detray::axis::regular<detray::container_types<vecmem::vector, detray::tuple, std::array, vecmem::jagged_vector, detray::dmap>, float> >, detray::axis::single_axis<detray::axis::circular<detray::axis::label::e_y>, detray::axis::regular<detray::container_types<vecmem::vector, detray::tuple, std::array, vecmem::jagged_vector, detray::dmap>, float> > >, detray::bins::single<detray::material_slab<float> >, detray::simple_serializer>]
/home/krasznaa/ATLAS/projects/detray/detray/core/include/detray/core/detail/container_buffers.hpp:67:77: note: no operand of the disjunction is satisfied
   67 |     std::derived_from<std::remove_cvref_t<T>, detray::detail::dbase_buffer> ||
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
   68 |     detray::detail::is_buffer_v<T>;
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                           
/home/krasznaa/ATLAS/projects/detray/detray/core/include/detray/core/detail/container_buffers.hpp:67:10: note: the operand ‘derived_from<typename std::remove_cvref<_Tp>::type, detray::detail::dbase_buffer>’ is unsatisfied because
   67 |     std::derived_from<std::remove_cvref_t<T>, detray::detail::dbase_buffer> ||
      |     ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   68 |     detray::detail::is_buffer_v<T>;
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/software/gcc/12.3.0/x86_64-el9/include/c++/12/concepts:67:13:   required for the satisfaction of ‘derived_from<typename std::remove_cvref<_Tp>::type, detray::detail::dbase_buffer>’ [with T = typename detray::detail::get_buffer<T, void>::type]
/home/krasznaa/ATLAS/projects/detray/detray/core/include/detray/core/detail/container_buffers.hpp:66:9:   required for the satisfaction of ‘device_buffer<typename detray::detail::get_buffer<T, void>::type>’ [with T = detray::grid_impl<detray::axis::multi_axis<false, detray::polar2D<detray::cmath<float> >, detray::axis::single_axis<detray::axis::closed<detray::axis::label::e_x>, detray::axis::regular<detray::container_types<vecmem::vector, detray::tuple, std::array, vecmem::jagged_vector, detray::dmap>, float> >, detray::axis::single_axis<detray::axis::circular<detray::axis::label::e_y>, detray::axis::regular<detray::container_types<vecmem::vector, detray::tuple, std::array, vecmem::jagged_vector, detray::dmap>, float> > >, detray::bins::single<detray::material_slab<float> >, detray::simple_serializer>]
...
```

Not impossible that GCC 12 is making a mistake here, but it certainly thinks that it would need to evaluate a concept to use that trait... :thinking: